File: common/src/main/java/com/pluxity/user/entity/CategoryBasedPermissionStrategy.java
package com.pluxity.user.entity;

import com.pluxity.global.annotation.ResolvePermission;
import java.util.Optional;

@ResolvePermission(PermissionType.CATEGORY)
public class CategoryBasedPermissionStrategy implements PermissionStrategy {
    @Override
    public boolean check(User user, Object resource) {
        return Optional.ofNullable(resource)
                .filter(r -> r instanceof Permissible)
                .map(r -> (Permissible) r)
                .map(p -> user.canAccess(p.getResourceType().getResourceName(), p.getResourceId()))
                .orElse(false);
    }
}

File: common/src/main/java/com/pluxity/user/entity/CategoryBasedPermissionStrategy.kt
package com.pluxity.user.entity

import com.pluxity.global.annotation.ResolvePermission

@ResolvePermission(PermissionType.CATEGORY)
class CategoryBasedPermissionStrategy : PermissionStrategy {
    override fun check(
        user: User,
        resource: Any,
    ): Boolean =
        (resource as? Permissible)?.let { permissible ->
            user.canAccess(permissible.resourceType.name, permissible.resourceId)
        } ?: false
}
